### ISSUE TITLE
data.table::update.dev.pkg() warning or error

### ISSUE THREAD

**Issue by drag05** [Rdatatable/data.table#4403]
`data.table v 1.12.8`:  

applying 
```
data.table::update.dev.pkg()
```
gives the following message in R v. 4.0.0, R Studio v 1. 2.5042 both with Admin rights:
```
R data.table package has been updated to NA (1.12.8)
Warning message:
In file.copy(savedcopy, lib, recursive = TRUE) :
  problem copying C:\R\R-4.0.0\library\00LOCK\data.table\libs\x64\datatable.dll to C:\R\R-4.0.0\library\data.table\libs\x64\datatable.dll: Permission denied
```

```
> sessionInfo()
R version 4.0.0 (2020-04-24)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_4.0.0    tools_4.0.0       data.table_1.12.8
```

**Comment by jangorecki** [comment]
It is actually the same problem as I have seen on CRAN windows checks. I was hoping it is a local CRAN issue related to new 4.0.0 setup. Thanks for reporting.

**Comment by tlapak** [comment]
I had the same issue on windows 10 using R 3.6.3. `update.dev.pkg()` downloads the binaries from gitlab, no? I don't have R tools installed on that machine. Specifying a `lib`, which I put in my documents folder, let me run the "update" successfully.

I suspect that the latest Windows update broke something.

**Comment by mattdowle** [comment]
It's a shame this is still happening without a better error message in R. We've seen it before in CRAN checks several times; the workaround there is that we ping Uwe and he reruns. CRAN checks are run in parallel. I don't know why data.table is reinstalled by CRAN each time but my guess is that when another package that uses data.table is running its checks and data.table's own checks are attempted to be run by CRAN (and an install happens first) it can't because the .dll is in use. Prior discussions with Uwe and others, if I remember correctly, are essentially that this is a Windows features. Windows locks its dll's. Unlike Linux and MacOS which have 

I don't think it is R 4.0.0 per se.  It's just that the event of R 4.0.0 triggers people to update who perhaps do not update very often.

Could the people who are having troubles ensure that all R sessions are shutdown, and then try upgrading/installing data.table again?  If you have a .RProfile, and that .RProfile loads data.table, then you have to turn that off otherwise data.table loads and the lock is opened on that dll.  This applies to every R package with a .dll as far as I know.

**Comment by jangorecki** [comment]
To install latest master please use this for now
```r
install.packages("data.table", repos="https://Rdatatable.gitlab.io/data.table")
```

**Comment by drag05** [comment]
@mattdowle 
R 4.0.0 asks for all previously R v < 4 installed packages to be re-installed. My issue popped up while I was reading about `fcase()` in News: the `data.table` had already been installed anew as requested ...

**Comment by drag05** [comment]
@jangorecki 

I have tried twice and got this every time:

```
> install.packages("data.table", repos="https://Rdatatable.gitlab.io/data.table")
trying URL 'https://Rdatatable.gitlab.io/data.table/bin/windows/contrib/4.0/data.table_1.12.9.zip'
Content type 'application/zip' length 2675878 bytes (2.6 MB)
downloaded 960 KB

Error in read.dcf(file.path(pkgname, "DESCRIPTION"), c("Package", "Type")) : 
  cannot open the connection
In addition: Warning messages:
1: In download.file(url, destfile, method, mode = "wb", ...) :
  downloaded length 983040 != reported length 2675878
2: In unzip(zipname, exdir = dest) : error 1 in extracting from zip file
3: In read.dcf(file.path(pkgname, "DESCRIPTION"), c("Package", "Type")) :
  cannot open compressed file 'data.table/DESCRIPTION', probable reason 'No such file or directory'
```
and

```
> sessionInfo()
R version 4.0.0 (2020-04-24)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_4.0.0 tools_4.0.0   
```

**Comment by jangorecki** [comment]
@drag05 install.packages warnings seems like issue with your connection. It doesn't download complete zip archive. AFAIK it is not related to the issue reported in this thread, and should be possible to fix by a retry (or more retries, maybe router restarting can help as well?).
Original issue has been addressed in PR #4404.

**Comment by drag05** [comment]
@jangorecki 

Thank you for the tip! I was inside my VPN. Now it works both ways - it seems:

```

> install.packages("data.table", repos="https://Rdatatable.gitlab.io/data.table")
trying URL 'https://Rdatatable.gitlab.io/data.table/bin/windows/contrib/4.0/data.table_1.12.9.zip'
Content type 'application/zip' length 2675878 bytes (2.6 MB)
downloaded 2.6 MB

package ‘data.table’ successfully unpacked and MD5 sums checked

The downloaded binary packages are in
        C:\Users\Dragos\AppData\Local\Temp\Rtmp2VNd5T\downloaded_packages
```
and

```
> data.table::update.dev.pkg()
R data.table package is up-to-date at dd7609e83132e19c9be80d71c73e8a8f95e19b27 (1.12.9)
```

So, I guess - no VPN allowed which may not be that good ...

**Comment by drag05** [comment]
Thank you for your prompt replies! I wonder if I should open a new issue involving VPN. 
Please advise!

**Comment by jangorecki** [comment]
I will re-open this issue, as I was able to reproduce the problem on windows. The issue was not just connection, but windows that locks dll, unloading namespace before reinstalling seems to remove the lock.

**Comment by drag05** [comment]
In my case the data.table wasn't loaded at any time, nor other R sessions open.
It could be that being inside a VPN triggered another issue.

**Comment by jangorecki** [comment]
when you call `data.table::` it automatically loads package namespace, do not confuse with attaching a namespace.

**Comment by drag05** [comment]
@jangorecki 
Now I see what you meant; if I understand you correctly, the functions after the `data.table::` should `unloadNamespace('data.table')` before doing everything else.

**Comment by jangorecki** [comment]
not function**s**, but this particular function, because it attempts to upgrade this namespace

**Comment by drag05** [comment]
I have miswrote. I meant it could happen to other namespaces too. Couldn't it?

**Comment by jangorecki** [comment]
It is related only to reinstalling pkg. Using `::` in general is fine, but when you upgrade pkg it is best to have its namespace unloaded.
